platform/src/main/java/org/stellar/anchor/platform/action/ActionHandler.java

```diff
@@ -0,0 +1,363 @@
+import org.stellar.anchor.sep24.Sep24Transaction;
+import org.stellar.anchor.sep24.Sep24TransactionStore;
+import org.stellar.anchor.sep31.Sep31Transaction;
+import org.stellar.anchor.sep31.Sep31TransactionStore;
+import org.stellar.anchor.util.AssetHelper;
+import org.stellar.anchor.util.GsonUtils;
+import org.stellar.anchor.util.SepHelper;
+import org.stellar.anchor.util.StringHelper;
+import org.stellar.sdk.AssetTypeCreditAlphaNum;
+import org.stellar.sdk.responses.AccountResponse;
+import org.stellar.sdk.responses.operations.PathPaymentBaseOperationResponse;
+import org.stellar.sdk.responses.operations.PaymentOperationResponse;
+
+public abstract class ActionHandler<T extends RpcActionParamsRequest> {
+
+  private static final Gson gson = GsonUtils.getInstance();
+
+  protected final Sep24TransactionStore txn24Store;
+  protected final Sep31TransactionStore txn31Store;
+  private final Validator validator;
+  private final Horizon horizon;
+  protected final AssetService assetService;
+  private final List<AssetInfo> assets;
+  private final Class<T> requestType;
+
+  public ActionHandler(
+      Sep24TransactionStore txn24Store,
+      Sep31TransactionStore txn31Store,
+      Validator validator,
+      Horizon horizon,
+      AssetService assetService,
+      Class<T> requestType) {
+    this.txn24Store = txn24Store;
+    this.txn31Store = txn31Store;
+    this.validator = validator;
+    this.horizon = horizon;
+    this.assetService = assetService;
+    this.assets = assetService.listAllAssets();
+    this.requestType = requestType;
+  }
+
+  @Transactional
+  public GetTransactionResponse handle(Object requestParams) throws AnchorException {
+    T request = gson.fromJson(gson.toJson(requestParams), requestType);
+    JdbcSepTransaction txn = getTransaction(request.getTransactionId());
+
+    if (txn == null) {
+      throw new InvalidRequestException(
+          String.format("Transaction with id[%s] is not found", request.getTransactionId()));
+    }
+
+    if (!getSupportedProtocols().contains(txn.getProtocol())) {
+      throw new InvalidRequestException(
+          String.format(
+              "Protocol[%s] is not supported by action[%s]", txn.getProtocol(), getActionType()));
+    }
+
+    if (!getSupportedStatuses(txn).contains(SepTransactionStatus.from(txn.getStatus()))) {
+      throw new InvalidRequestException(
+          String.format(
+              "Action[%s] is not supported for status[%s]", getActionType(), txn.getStatus()));
+    }
+
+    updateTransaction(txn, request);
+
+    return toGetTransactionResponse(txn, assetService);
+  }
+
+  public abstract ActionMethod getActionType();
+
+  protected abstract SepTransactionStatus getNextStatus(JdbcSepTransaction txn, T request)
+      throws InvalidRequestException, InvalidParamsException;
+
+  protected abstract Set<SepTransactionStatus> getSupportedStatuses(JdbcSepTransaction txn);
+
+  protected abstract Set<String> getSupportedProtocols();
```

Comment:
I updated code to use enum, but i removed `getSupportedProtocols` method and moved this validation to `getSupportedStatuses`, because we still will need to duplicate protocol validation when we add support for SEP-31 etc. Also updated exception message